crates/sui-json-rpc/src/metrics.rs

```diff
@@ -156,10 +156,30 @@ where
                         .req_latency_by_route
                         .with_label_values(&[&name])
                         .observe(req_latency_secs);
-
-                    if res.status().is_server_error() {
-                        metrics.errors_by_route.with_label_values(&[&name]).inc();
+                    // Parse error code from response
+                    #[derive(Deserialize)]
+                    struct RPCResponse {
+                        #[serde(default)]
+                        error: Option<RPCError>,
+                    }
+                    #[derive(Deserialize)]
+                    struct RPCError {
+                        code: i16,
+                    }
+                    let (parts, body) = res.into_parts();
+                    let bytes = body::to_bytes(body).await?;
```

Comment:
IIUC, this line can cause the request to fail if at this point, we can't convert the body to bytes.  I'd prefer to let the request go through in that case, but with an unknown error code.

Is there a way to parse the bytes from the response body without consuming the initial `Response` you got?  That seems like the first stumbling block.

Reply:
if this fail, the body will be lost because `body::to_bytes` owns the body and there is nothing I can return except error...

Sadly the response and body object is not clone-able, and I think there are no way to peek/borrow the body content without consuming the response.

Reply:
Are you sure?  I see an accessor for `body` by reference when I check the implementations on `Response` in my editor -- it's [this one](https://docs.rs/hyper/latest/hyper/struct.Response.html#method.body) in the docs.

Reply:
ya pretty sure, the only data we can get from the data reference is the size hint and `is_end_stream`, the Body will be consumed when you access the data content `fn data(&mut self)`

Reply:
😞 sad times, yeah you're right.  Okay, we'll live with it!  Resolving.